gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/api/VerifyApiPathDomainServiceImpl.java

```diff
@@ -0,0 +1,283 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.apim.infra.domain_service.api;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Strings;
+import com.google.common.net.InternetDomainName;
+import io.gravitee.apim.core.api.domain_service.VerifyApiPathDomainService;
+import io.gravitee.apim.core.api.model.Path;
+import io.gravitee.definition.model.DefinitionVersion;
+import io.gravitee.definition.model.v4.listener.http.HttpListener;
+import io.gravitee.repository.exceptions.TechnicalException;
+import io.gravitee.repository.management.api.ApiRepository;
+import io.gravitee.repository.management.api.EnvironmentRepository;
+import io.gravitee.repository.management.api.search.ApiCriteria;
+import io.gravitee.repository.management.api.search.ApiFieldFilter;
+import io.gravitee.repository.management.model.Api;
+import io.gravitee.rest.api.service.common.ExecutionContext;
+import io.gravitee.rest.api.service.exceptions.EnvironmentNotFoundException;
+import io.gravitee.rest.api.service.exceptions.TechnicalManagementException;
+import io.gravitee.rest.api.service.v4.exception.HttpListenerPathMissingException;
+import io.gravitee.rest.api.service.v4.exception.InvalidHostException;
+import io.gravitee.rest.api.service.v4.exception.InvalidPathException;
+import io.gravitee.rest.api.service.v4.exception.PathAlreadyExistsException;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class VerifyApiPathDomainServiceImpl implements VerifyApiPathDomainService {
+
+    private static final Pattern DUPLICATE_SLASH_REMOVER = Pattern.compile("[//]+");
+    private static final Pattern VALID_PATH = Pattern.compile("^[/.a-zA-Z0-9-_]*$");
+    private static final String URI_PATH_SEPARATOR = "/";
+    private static final char URI_PATH_SEPARATOR_CHAR = '/';
+
+    private final EnvironmentRepository environmentRepository;
+    private final ApiRepository apiRepository;
+    private final ObjectMapper objectMapper;
+
+    public VerifyApiPathDomainServiceImpl(
+        @Lazy final EnvironmentRepository environmentRepository,
+        @Lazy final ApiRepository apiRepository,
+        final ObjectMapper objectMapper
+    ) {
+        this.environmentRepository = environmentRepository;
+        this.apiRepository = apiRepository;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public List<Path> verifyApiPaths(ExecutionContext executionContext, String apiId, List<Path> paths) {
```

Comment:
I think I would have created an `ApiPaths` object holding the list of path and providing methods to validate and sanitize internally. wdyt ?

Reply:
Sanitizing requires external information (like domainRestrictions coming from environment) si we can not keep this logic in a simple object.